File: src/Supply.cpp
#include "Supply.hpp"
#include "Warehouse.hpp"
#include <iostream>
using namespace std;

int Supply::nextSupplyId = 1;

Supply::Supply(const string& supplierName, const string& supplyDate): supplyId(nextSupplyId++), supplierName(supplierName), supplyDate(supplyDate), status("Новая") {}

bool Supply::addProduct(Product* product, int quantity) {
    if (product && quantity > 0) {
        products.push_back(make_pair(product, quantity));
    cout << "Добавлен товар '" << product->getType() << "' x" << quantity<< " в поставку #" << supplyId << endl;
        return true;
    }
    return false;
}

bool Supply::removeProduct(int productId) {
    for (auto it = products.begin(); it != products.end(); ++it) {
        if (it->first->getProductId() == productId) {
            cout << "Удалён товар '" << it->first->getType() << "' из поставки" << endl;
            products.erase(it);
            return true;
        }
    }
    return false;
}

void Supply::processSupply(Warehouse* warehouse) {
    cout << "\n=== Обработка поставки #" << supplyId << " ===" << endl;
    cout << "Поставщик: " << supplierName << endl;
    cout << "Дата: " << supplyDate << endl;

    for (const auto& item : products) {
        Product* product = warehouse->findProduct(item.first->getProductId());
        if (product) {
            product->incQuantity(item.second);
            cout << "Товар '" << product->getType() << "' пополнен на "<< item.second << " на складе" << endl;
        }
    }

    status = "Обработана";
    cout << "Поставка #" << supplyId << " успешно обработана" << endl;
}

string Supply::getSupplyInfo() const {
    string s_inf = "Поставка #" + to_string(supplyId) + " от " + supplyDate +"\nПоставщик: " + supplierName +"\nСтатус: " + status + "\nТовары:";
    for (const auto& item : products) {
        s_inf += "\n- " + item.first->getType() + " x" + to_string(item.second);
    }
    return s_inf;
}

void Supply::displaySupply() const {
    cout << getSupplyInfo() << endl;
}

int Supply::getSupplyId() const { 
    return supplyId;
}
string Supply::getSupplierName() const {
    return supplierName;
}
string Supply::getSupplyDate() const { 
    return supplyDate;
}
string Supply::getStatus() const {
    return status; 
}

File: warehouse/Supply.java
package com.warehouse;

import java.util.ArrayList;
import java.util.List;

public class Supply {
    private static int nextSupplyId = 1;
    
    private int supplyId;
    private List<Product> products;
    private List<Integer> quantities;
    private String supplierName;
    private String supplyDate;
    private String status;
    
    public Supply(String supplierName, String supplyDate) {
        this.supplyId = nextSupplyId++;
        this.supplierName = supplierName;
        this.supplyDate = supplyDate;
        this.products = new ArrayList<>();
        this.quantities = new ArrayList<>();
        this.status = "Создана";
    }
    
    public boolean addProduct(Product product, int quantity) {
        if (product != null && quantity > 0) {
            products.add(product);
            quantities.add(quantity);
            System.out.println("Добавлен товар '" + product.getType() + "' x" + quantity + " в поставку #" + supplyId);
            return true;
        }
        return false;
    }
    
    public boolean removeProduct(int productId) {
        for (int i = 0; i < products.size(); i++) {
            if (products.get(i).getProductId() == productId) {
                products.remove(i);
                quantities.remove(i);
                return true;
            }
        }
        return false;
    }
    
    public void processSupply(Warehouse warehouse) {
        System.out.println("\n=== Обработка поставки #" + supplyId + " ===");
        System.out.println("Поставщик: " + supplierName);
        System.out.println("Дата: " + supplyDate);
        
        for (int i = 0; i < products.size(); i++) {
            Product product = products.get(i);
            int quantity = quantities.get(i);
            
            Product warehouseProduct = warehouse.findProduct(product.getProductId());
            if (warehouseProduct != null) {
                warehouseProduct.addQuantity(quantity);
                System.out.println("Товар '" + product.getType() + "' добавлен в количестве " + quantity + " на склад");
            } else {
                Product newProduct = new Product(product.getProductId(), product.getType(),  quantity, product.getDescription(), product.getPrice());warehouse.addProduct(newProduct);
            }
        }
        
        status = "Обработана";
        System.out.println("Поставка #" + supplyId + " успешно обработана");
    }
    
    public void displaySupply() {
        System.out.println("Поставка #" + supplyId + " от " + supplyDate);
        System.out.println("Поставщик: " + supplierName);
        System.out.println("Статус: " + status);
        System.out.println("Товары:");
        for (int i = 0; i < products.size(); i++) {
            System.out.println("- " + products.get(i).getType() + " x" + quantities.get(i));
        }
    }
    
    public int getSupplyId() { 
        return supplyId;
     }
    public String getSupplierName() {
         return supplierName; 
        }
    public String getSupplyDate() { 
        return supplyDate; 
    }
    public String getStatus() { 
        return status; 
    }
    public String getSupplyInfo() {
        return "Поставка #" + supplyId + " от " + supplyDate + ", Поставщик: " + supplierName + ", Статус: " + status;
    }
}